### PR TITLE
Spark, Flink: Add null engineSchema fallback for format model writers

### PR DESCRIPTION
**Pull request by joyhaldar**
When `engineSchema` is not set, format model writers now derive it from the Iceberg schema instead of failing with a null error.

[Discussed](https://github.com/apache/iceberg/pull/15441#discussion_r2852437765) in [Add TCK for File Format API](https://github.com/apache/iceberg/pull/15441).

**Spark (4.1, 4.0, 3.5, 3.4):** Added `SparkAvroWriter` constructor with null fallback. Updated `SparkFormatModels`. Parquet already had this, ORC ignores `engineSchema`.

**Flink (2.1, 2.0, 1.20):** Added null fallback in `FlinkAvroWriter`, `FlinkParquetWriters`, and `FlinkOrcWriter`. Updated `FlinkFormatModels`.

**Tests:** Added `testDataWriterEngineWriteWithoutEngineSchema` and `testEqualityDeleteWriterEngineWriteWithoutEngineSchema` in `BaseFormatModelTests` to cover the null `engineSchema` path. Existing tests with explicit `engineSchema` are kept.

Part of #15415

### CODE REVIEW

File: spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/SparkAvroWriter.java [hunk 1]
```diff
@@ -46,6 +47,12 @@ public SparkAvroWriter(StructType dsSchema) {
     this.dsSchema = dsSchema;
   }
 
+  public static SparkAvroWriter buildWriter(
```
Comment (pvary) [inline spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/SparkAvroWriter.java]: Why not a new constructor, like:
```
  public SparkAvroWriter(org.apache.iceberg.Schema icebergSchema, StructType dsSchema) {
    this.dsSchema =  dsSchema != null ? dsSchema : SparkSchemaUtil.convert(icebergSchema);
  }
```

File: flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetWriters.java [hunk 2]
```diff
@@ -80,6 +81,12 @@
 public class FlinkParquetWriters {
   private FlinkParquetWriters() {}
 
+  public static <T> ParquetValueWriter<T> buildWriter(
+      org.apache.iceberg.Schema icebergSchema, MessageType type, RowType engineSchema) {
```
Comment (pvary) [inline flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetWriters.java]: nit: import org.apache.iceberg.Schema

File: flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkAvroWriter.java [hunk 3]
```diff
@@ -43,6 +44,12 @@ public FlinkAvroWriter(RowType rowType) {
     this.rowType = rowType;
   }
 
+  public static FlinkAvroWriter buildWriter(
+      org.apache.iceberg.Schema icebergSchema, RowType engineSchema) {
```
Comment (pvary) [inline flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkAvroWriter.java]: import

**Comment by pvary** [comment]
Update the tests to remove the unnecessary engineSchema settings

**Comment by joyhaldar** [comment]
> Update the tests to remove the unnecessary engineSchema settings

Removing `.engineSchema()` from `BaseFormatModelTests` will also break older Spark and Flink versions that don't have the fallback yet. Should I wait until backports are done?

File: spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/SparkAvroWriter.java (hunk 1 above)
Comment (joyhaldar) [reply spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/SparkAvroWriter.java]: Sorry about that, changed to constructor.

File: flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetWriters.java (hunk 2 above)
Comment (joyhaldar) [reply flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetWriters.java]: Sorry about that, added import.

File: flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkAvroWriter.java (hunk 3 above)
Comment (joyhaldar) [reply flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkAvroWriter.java]: Can't import `org.apache.iceberg.Schema`, as `org.apache.avro.Schema` is [already imported](https://github.com/apache/iceberg/blob/1009ef41b25b4602baa949c15c31db34b7a2ad96/flink/v2.1/flink/src/main/java/org/apache/iceberg/flink/data/FlinkAvroWriter.java#L27). Is fully qualified okay?

File: spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/SparkAvroWriter.java [hunk 4]
```diff
@@ -46,6 +47,10 @@ public SparkAvroWriter(StructType dsSchema) {
     this.dsSchema = dsSchema;
   }
 
+  public SparkAvroWriter(org.apache.iceberg.Schema icebergSchema, StructType engineSchema) {
+    this(engineSchema != null ? engineSchema : SparkSchemaUtil.convert(icebergSchema));
```
Comment (huaxingao) [inline spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/SparkAvroWriter.java]: Do we have tests to cover both the null and not null path?

**Comment by pvary** [comment]
> Removing `.engineSchema()` from `BaseFormatModelTests` will also break older Spark and Flink versions that don't have the fallback yet. Should I wait until backports are done?

We can update all of the Spark and Flink versions in one PR, and then we can have a test where it is removed.

Also, reflecting on @huaxingao’s comment, I think we should keep the other tests as well.

File: spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/SparkAvroWriter.java (hunk 4 above)
Comment (joyhaldar) [reply spark/v4.1/spark/src/main/java/org/apache/iceberg/spark/data/SparkAvroWriter.java]: Thank you for the review @huaxingao. 

I have added two new tests in `BaseFormatModelTests` that cover the null path based on your and @pvary's suggestions, `testDataWriterEngineWriteWithoutEngineSchema` and `testEqualityDeleteWriterEngineWriteWithoutEngineSchema`. The existing tests with explicit `engineSchema` are also present.

**Comment by joyhaldar** [comment]
> > Removing `.engineSchema()` from `BaseFormatModelTests` will also break older Spark and Flink versions that don't have the fallback yet. Should I wait until backports are done?
> 
> We can update all of the Spark and Flink versions in one PR, and then we can have a test where it is removed.
> 
> Also, reflecting on @huaxingao’s comment, I think we should keep the other tests as well.

Thank you Péter. I have applied the fallback to all Spark and Flink versions. I have also added new tests for the null engineSchema path while keeping existing tests that set it explicitly.

File: data/src/test/java/org/apache/iceberg/data/BaseFormatModelTests.java [hunk 5]
```diff
@@ -140,6 +140,45 @@ void testDataWriterEngineWriteGenericRead(FileFormat fileFormat, DataGenerator d
     DataTestHelpers.assertEquals(schema.asStruct(), genericRecords, readRecords);
   }
 
+  @ParameterizedTest
+  @FieldSource("FORMAT_AND_GENERATOR")
+  /** Write with engine type T without explicit engineSchema, read with Generic Record */
```
Comment (pvary) [inline data/src/test/java/org/apache/iceberg/data/BaseFormatModelTests.java]: The comments and annotations should be reordered so that this comment can be treated as Javadoc.

```suggestion
  /** Write with engine type T without explicit engineSchema, read with Generic Record */
  @ParameterizedTest
  @FieldSource("FORMAT_AND_GENERATOR")
```

File: data/src/test/java/org/apache/iceberg/data/BaseFormatModelTests.java [hunk 6]
```diff
@@ -225,6 +264,54 @@ void testEqualityDeleteWriterEngineWriteGenericRead(
     DataTestHelpers.assertEquals(schema.asStruct(), genericRecords, readRecords);
   }
 
+  @ParameterizedTest
+  @FieldSource("FORMAT_AND_GENERATOR")
+  /**
+   * Write equality deletes with engine type T without explicit engineSchema, read with Generic
+   * Record
+   */
```
Comment (pvary) [inline data/src/test/java/org/apache/iceberg/data/BaseFormatModelTests.java]:
```suggestion
  /**
   * Write equality deletes with engine type T without explicit engineSchema, read with Generic
   * Record
   */
  @ParameterizedTest
  @FieldSource("FORMAT_AND_GENERATOR")
```

File: data/src/test/java/org/apache/iceberg/data/BaseFormatModelTests.java [hunk 7]
```diff
@@ -140,6 +140,45 @@ void testDataWriterEngineWriteGenericRead(FileFormat fileFormat, DataGenerator d
...
+    InputFile inputFile = encryptedFile.encryptingOutputFile().toInputFile();
+    List<Record> readRecords;
+    try (CloseableIterable<Record> reader =
+        FormatModelRegistry.readBuilder(fileFormat, Record.class, inputFile)
+            .project(schema)
+            .build()) {
+      readRecords = ImmutableList.copyOf(reader);
+    }
+
+    DataTestHelpers.assertEquals(schema.asStruct(), genericRecords, readRecords);
```
Comment (pvary) [inline data/src/test/java/org/apache/iceberg/data/BaseFormatModelTests.java]: Could we generalize this to a method?
I think @Guosmilesmile already did this in his #15633 PR

Comment (joyhaldar) [reply data/src/test/java/org/apache/iceberg/data/BaseFormatModelTests.java]: I have extracted `readAndAssertGenericRecords` helper. Inspired by @Guosmilesmile's `writeGenericRecords` pattern in #15633.

**Comment by pvary** [comment]
Merged to main.
Thanks @joyhaldar for the PR and @huaxingao for the reivew!
